Fix location state (#394)

BlazPocrnja opened the pull request:
Closes #394. This change deep compares location state objects to prevent incorrectly pushing to history multiple times when state is changed.

- [x] Ran all tests

supasate commented on src/ConnectedRouter.js:
```diff
@@ -2,6 +2,7 @@ import React, { PureComponent } from 'react'
 import PropTypes from 'prop-types'
 import { connect, ReactReduxContext } from 'react-redux'
 import { Router } from 'react-router'
+import { isEqual } from 'lodash'
```
`import isEqual from 'lodash.isequal'` to avoid loading the whole lodash module. 

Please run `yarn add -P lodash.isequal` to update package.json and yarn.lock as peerDependencies.

supasate commented:
Hi @BlazPocrnja , would you mind adding a unit test for this behavior?

BlazPocrnja commented:
@supasate I added unit tests and changed the lodash function to isEqualWith to allow for custom comparator function as per this comment https://github.com/supasate/connected-react-router/issues/394#issuecomment-597730704.

supasate commented on src/ConnectedRouter.js:
```diff
@@ -37,15 +38,15 @@ const createConnectedRouter = (structure) => {
           hash: hashInHistory,
           state: stateInHistory,
-        } = history.location
+				} = history.location
```
nit: use spaces, not tabs.

supasate commented on src/ConnectedRouter.js:
```diff
@@ -104,7 +105,8 @@ const createConnectedRouter = (structure) => {
     basename: PropTypes.string,
     children: PropTypes.oneOfType([ PropTypes.func, PropTypes.node ]),
-    onLocationChanged: PropTypes.func.isRequired,
+		onLocationChanged: PropTypes.func.isRequired,
```
ditto: spaces

supasate commented on src/ConnectedRouter.js:
```diff
@@ -104,7 +105,8 @@ const createConnectedRouter = (structure) => {
     children: PropTypes.oneOfType([ PropTypes.func, PropTypes.node ]),
-    onLocationChanged: PropTypes.func.isRequired,
+		onLocationChanged: PropTypes.func.isRequired,
+		stateCompareFunction: PropTypes.func,
```
ditto: spaces

supasate commented on src/ConnectedRouter.js:
```diff
@@ -18,7 +19,7 @@ const createConnectedRouter = (structure) => {
       super(props)
 
-      const { store, history, onLocationChanged } = props
+      const { store, history, onLocationChanged, stateCompareFunction} = props
```
nit: add a space before the closing curling brace.

supasate commented on test/ConnectedRouter.test.js:
```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
+          },
+          action: 'PUSH',
+        }
+			})
```
ditto: spaces

supasate commented on test/ConnectedRouter.test.js:
```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
+        }
+			})
+			
+			expect(props.history.entries).toHaveLength(3)
```
ditto: spaces

supasate replied:
ditto: spaces

supasate commented on test/ConnectedRouter.test.js:
```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
+        }
+			})
+			
+			expect(props.history.entries).toHaveLength(3)
```
ditto

supasate commented on test/ConnectedRouter.test.js:
```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
+      })
+
+      expect(props.history.entries).toHaveLength(4)
+    })
```
Can you split this into another test? So, we have one for testing that it doesn't update if the location state structure doesn't change, and another one for testing that it updates if the location state structure changes.

supasate commented on test/ConnectedRouter.test.js:
```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
+          },
+          action: 'PUSH',
+        }
+			})
```
ditto

supasate commented on test/ConnectedRouter.test.js:
```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
+							return a === undefined  || (a.foo === "baz" && b.foo === 'bar') ? true : false
+            }}
+            {...props}
+					>
```
ditto

supasate commented on test/ConnectedRouter.test.js:
```diff
@@ -135,6 +135,126 @@ describe('ConnectedRouter', () => {
+        <Provider store={store}>
+          <ConnectedRouter
+            stateCompareFunction={(a, b) => {
+							return a === undefined  || (a.foo === "baz" && b.foo === 'bar') ? true : false
```
No need to add `a === undefined` which we doesn't test it. Only the second comparison is enough to test this comparison feature.

```suggestion
              return (a.foo === 'baz' && b.foo === 'bar') ? true : false
```

BlazPocrnja commented:
@supasate Sorry for the delay! I made the changes you requested. Except when testing the custom state comparator function, there needs to be a check when the state's are undefined (which is the case when you first push to history, or the store). I rewrote the test to make it a bit clearer. It's rebased now too!

supasate commented:
LGTM. Thanks for your contribution @BlazPocrnja !

supasate commented:
I just noticed that it breaks the test because lodash.isequalwith is not listed in devDependencies. (https://travis-ci.org/github/supasate/connected-react-router/builds/665324580?utm_medium=notification&utm_source=github_status). I'm not sure why Travis didn't show up here. I'm fixing in #408.

BlazPocrnja commented:
My apologies! Looks like it's passing now, so that's a relief. Thanks!
